internal/store/memory/store.go

```diff
@@ -177,56 +177,46 @@ func (s *Store) DeleteRoute(ctx context.Context, id string) error {
 	return s.Sync(ctx)
 }
 
-func (s *Store) UpsertRoute(ctx context.Context, route store.Route) error {
+func (s *Store) UpsertRoute(ctx context.Context, route store.Route, updateConditionFn func(current store.Route) bool) error {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
 	id := route.ID()
 
-	switch compareRoutes(s.routes[id], route) {
-	case store.CompareResultInvalid, store.CompareResultNewer:
-		// we have an old or invalid route, ignore it
+	if updateConditionFn != nil && !updateConditionFn(s.routes[id]) {
+		// No-op
 		return nil
-	case store.CompareResultNotEqual:
-		s.logger.Trace("detected route state change", "id", id)
-		s.routes[id] = route
+	}
 
-		// bind to gateways
-		for _, gateway := range s.gateways {
-			gateway.TryBind(ctx, route)
-		}
+	s.routes[id] = route
+
+	// bind to gateways
+	for _, gateway := range s.gateways {
+		gateway.TryBind(ctx, route)
 	}
 
 	// sync the gateways to consul and route statuses to k8s
 	return s.Sync(ctx)
 }
 
-func compareRoutes(a, b store.Route) store.CompareResult {
-	if b == nil {
-		return store.CompareResultInvalid
-	}
-	if a == nil {
-		return store.CompareResultNotEqual
-	}
-	return a.Compare(b)
-}
-
-func (s *Store) UpsertGateway(ctx context.Context, gateway store.Gateway) error {
+func (s *Store) UpsertGateway(ctx context.Context, gateway store.Gateway, updateConditionFn func(current store.Gateway) bool) error {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
 	id := gateway.ID()
 
 	current, found := s.gateways[id]
+	var currentGW store.Gateway
+	if found {
+		currentGW = current.Gateway
+	}
 
-	if !current.ShouldUpdate(gateway) {
```

Comment:
We previously assumed `current` was non-nil. Why do we now assume it might be nil?

Reply:
We hadn't assumed it was non-nil before, the logic was just hidden inside of the `ShouldUpdate` function on the `K8sGateway` structure:

https://github.com/hashicorp/consul-api-gateway/blob/564ad2b0114407168dfd0e9066bd463fd61a9efa/internal/k8s/reconciler/gateway.go#L352-L359